File: Backend/ProjectK.Backend/ProjectK.Common/Models/Enums/PlastLadder.cs
namespace ProjectK.Common.Models.Enums
{
    /// <summary>
    /// The order a person passes the ступені in, and which of them a kurin of a given branch
    /// normally records. Order lives here rather than in the enum because the enum's numbers are
    /// storage and cannot be reordered — see <see cref="PlastLevel"/>.
    /// </summary>
    public static class PlastLadder
    {
        /// <summary>Every ступінь, in the order it is reached.</summary>
        public static readonly IReadOnlyList<PlastLevel> InOrder =
        [
            PlastLevel.Entry,
            PlastLevel.Prykhylnyk,
            PlastLevel.Uchasnyk,
            PlastLevel.Rozviduvach,
            PlastLevel.Skob,
            PlastLevel.HetmanskiySkob,
            PlastLevel.Starshoplastun,
            PlastLevel.Senior,
            PlastLevel.SeniorPratsi,
            PlastLevel.SeniorDovirja,
            PlastLevel.SeniorKerivnytstva
        ];

        private static readonly IReadOnlyList<PlastLevel> ThroughYouth =
        [
            PlastLevel.Entry,
            PlastLevel.Prykhylnyk,
            PlastLevel.Uchasnyk,
            PlastLevel.Rozviduvach,
            PlastLevel.Skob,
            PlastLevel.HetmanskiySkob
        ];

        private static readonly IReadOnlyList<PlastLevel> ThroughSenior =
        [
            .. ThroughYouth,
            PlastLevel.Starshoplastun
        ];

        /// <summary>
        /// What a kurin of this branch is shown by default. Cumulative on purpose: a УПС kurin still
        /// wants to see when its people were заприсяжені. This is a **default**, not a restriction —
        /// the провід turns individual columns on and off from there.
        /// </summary>
        public static IReadOnlyList<PlastLevel> DefaultFor(KurinBranch branch)
            => branch switch
            {
                KurinBranch.USP => ThroughSenior,
                KurinBranch.UPS => InOrder,
                _ => ThroughYouth
            };
    }
}

File: Frontend/projectk-frontend/src/app/features/kurinModule/common/models/enums/plast-ladder.ts
import { KurinBranch } from './kurin-branch.enum';
import { PlastLevel } from './plast-level.enum';

/**
 * Порядок, у якому людина проходить ступені, і що з них курінь кожної гілки звично записує.
 *
 * Порядок живе тут, а не в самому enum: ступінь зберігається числом, і числа заморожені —
 * `Prykhylnyk` дописаний останнім, хоч на драбині стоїть другим. Дзеркалить `PlastLadder` на бекенді.
 */
export const PLAST_LADDER: readonly PlastLevel[] = [
  PlastLevel.Entry,
  PlastLevel.Prykhylnyk,
  PlastLevel.Uchasnyk,
  PlastLevel.Rozviduvach,
  PlastLevel.Skob,
  PlastLevel.HetmanskiySkob,
  PlastLevel.Starshoplastun,
  PlastLevel.Senior,
  PlastLevel.SeniorPratsi,
  PlastLevel.SeniorDovirja,
  PlastLevel.SeniorKerivnytstva
];

/**
 * Підпис ступеня — так, як його пишуть у звітах. Дата вступу до Пласту стоїть під «пл. неім.»: це і
 * є ступінь пластуна неіменованого, а не окремий службовий рядок перед драбиною.
 */
export const PLAST_LEVEL_COLUMN_LABELS: Record<PlastLevel, string> = {
  [PlastLevel.Entry]: 'пл. неім.',
  [PlastLevel.Prykhylnyk]: 'пл. прих.',
  [PlastLevel.Uchasnyk]: 'пл. уч.',
  [PlastLevel.Rozviduvach]: 'пл. розв.',
  [PlastLevel.Skob]: 'пл. скоб / вірл.',
  [PlastLevel.HetmanskiySkob]: 'пл. гетьм. скоб / вірл.',
  [PlastLevel.Starshoplastun]: 'Старшопластун',
  [PlastLevel.Senior]: 'Перехід в УПС',
  [PlastLevel.SeniorPratsi]: 'Сен. праці',
  [PlastLevel.SeniorDovirja]: "Сен. довір'я",
  [PlastLevel.SeniorKerivnytstva]: 'Сен. керівництва'
};

const THROUGH_YOUTH: readonly PlastLevel[] = PLAST_LADDER.slice(0, 6);
const THROUGH_SENIOR: readonly PlastLevel[] = PLAST_LADDER.slice(0, 7);

/**
 * Які ступені показувати куреню цієї гілки **за замовчуванням**. Накопичувально: УПС-курінь однаково
 * хоче бачити, коли його люди були заприсяжені. Це дефолт, а не обмеження — далі провід вмикає й
 * вимикає колонки сам.
 */
export function defaultLevelsFor(branch: KurinBranch | null | undefined): readonly PlastLevel[] {
  switch (branch) {
    case KurinBranch.USP:
      return THROUGH_SENIOR;
    case KurinBranch.UPS:
      return PLAST_LADDER;
    default:
      return THROUGH_YOUTH;
  }
}
